src/decision-engine/index.js

```diff
@@ -1,110 +1,126 @@
 'use strict'
 
-const debounce = require('just-debounce-it')
+const CID = require('cids')
 
 const Message = require('../types/message')
+const WantType = Message.WantType
 const Wantlist = require('../types/wantlist')
 const Ledger = require('./ledger')
-const { logger, groupBy, pullAllWith, uniqWith } = require('../utils')
+const RequestQueue = require('./req-queue')
+const TaskMerger = require('./task-merger')
+const { logger } = require('../utils')
 
-const MAX_MESSAGE_SIZE = 512 * 1024
+// The ideal size of the batched payload. We try to pop this much data off the
+// request queue, but it may be a little more or less depending on what's in
+// the queue.
+const TARGET_MESSAGE_SIZE = 16 * 1024
+
+// The maximum size of the block in bytes up to which we will replace a
+// want-have with a want-block
+const MAX_BLOCK_SIZE_REPLACE_HAS_WITH_BLOCK = 1024
 
 class DecisionEngine {
-  constructor (peerId, blockstore, network, stats) {
+  constructor (peerId, blockstore, network, stats, opts) {
     this._log = logger(peerId, 'engine')
     this.blockstore = blockstore
     this.network = network
     this._stats = stats
+    this._opts = this._processOpts(opts)
 
     // A list of of ledgers by their partner id
     this.ledgerMap = new Map()
     this._running = false
 
-    // List of tasks to be processed
-    this._tasks = []
+    // Queue of want-have / want-block per peer
+    this._requestQueue = new RequestQueue(TaskMerger)
+  }
 
-    this._outbox = debounce(this._processTasks.bind(this), 100)
+  _processOpts (opts) {
+    return {
+      ...{
+        maxBlockSizeReplaceHasWithBlock: MAX_BLOCK_SIZE_REPLACE_HAS_WITH_BLOCK,
+        targetMessageSize: TARGET_MESSAGE_SIZE
+      },
+      ...opts
+    }
   }
 
-  async _sendBlocks (peer, blocks) {
-    // split into messages of max 512 * 1024 bytes
-    let total = blocks.reduce((acc, b) => {
-      return acc + b.data.byteLength
-    }, 0)
+  _outbox () {
+    setImmediate(() => this._processTasks())
+  }
 
-    // If the blocks fit into one message, send the message right away
-    if (total < MAX_MESSAGE_SIZE) {
-      await this._sendSafeBlocks(peer, blocks, 0)
+  // Pull tasks off the request queue and send a message to the corresponding
+  // peer
+  async _processTasks () {
+    if (!this._running) {
```

Comment:
> Should this be if running return, not if not running return?

When the engine is stopped, `this._running` is set to `true`, so we want to bail out of the `processTasks` function.

> This creates unbounded concurrency to the size of our task list

IIUC Javascript's thread model will queue up multiple setImmediate() calls to execute in series, on the next tick. The first thing `processTasks()` does is pop tasks off the queue, and if there are no tasks it bails out. So as the code stands it will always have at least one thread per message, and any further threads will exit after popping from an empty queue.

I agree that it would be conceptually simpler to use a set of workers, and even better if it's a dynamic set, starting at one worker and then growing and shrinking as needed, so as to use less resources. Should we tackle that in a subsequent PR?

Reply:
> IIUC Javascript's thread model will queue up multiple setImmediate() calls to execute in series, on the next tick

I don't mean literal concurrency since we only have 1 thread to begin with. I am referring too the fact that we have multiple points of entry for adding `processTasks` to the queue. It shouldn't be terribly harmful, but we may end up with a scenario where we have to call `processTasks` several times to drain the call stack even though there is no more tasks to be run. But this can be tackled in a separate PR is part of a worker flow improvement.

> Should we tackle that in a subsequent PR?

Yeah I think tackling this in a followup PR of the flow would be good to avoid tacking too much onto this.